### PR TITLE
Request privileges to replace windows

### PR DESCRIPTION
**Pull request by Lambourl**
This Pr is intended to request the elevation of privilege for the replacement of windows

### CODE REVIEW

File: .nxxm/deps [hunk 1]
```diff
@@ -4,6 +4,6 @@
    , "cpp-pre/file" : {}
    , "nxxm/cli_widgets" : { "@" : "master" }
    , "kuba--/zip" : { "@" : "v0.1.9"  }
-   , "daminetreg/xxhr" : { "@" : "feature/nxxm-migration" }
+   , "nxxm/xxhr" : { "@" : "feature/fix-decoding-nullptr-ending-buffer" }
```
Comment (pysco68) [inline .nxxm/deps]: Is that the same as using in nxxm-src?

File: src/shipxx/shipxx.hxx [hunk 2]
```diff
@@ -19,7 +19,7 @@ namespace shipxx {
   namespace detail {
     inline void extract(fs::path file, fs::path out) {
       auto on_extract_entry = [](const char *filename, void *) {
-        log::debug() << "extracted " << filename << std::endl;
+        std::cout << "extracted " << filename << std::endl;
```
Comment (pysco68) [inline src/shipxx/shipxx.hxx:22]: I guess you should revert all those to using ``log::debug()`

File: src/upgrd/upgrd.hxx [hunk 3]
```diff
@@ -206,17 +206,17 @@ namespace upgrd {
...
 
                     _log << "We will close, the next time you will come back you will be in the newer version" << std::endl;
   
                     #if BOOST_OS_WINDOWS
-                    auto str_cmd = "timeout /t 3 /nobreak & del /F /Q "s + _app_path.make_preferred().string() + " & "
-                      + "move /Y " + upgraded_app.make_preferred().string() + " " + _app_path.make_preferred().string();
-
-                    bp::spawn(system_shell, "/c", str_cmd.data());
-
+                    auto system_shell = bp::search_path("powershell.exe");
```
Comment (pysco68) [inline src/upgrd/upgrd.hxx:215]: We should add a second message on windows re. the potentially upcoming UAC prompt.

"Tipi may need administrative privileges to apply the update."

**Review by pysco68 — CHANGES_REQUESTED** [review]

File: .nxxm/deps (hunk 1 above)
Comment (Lambourl) [reply .nxxm/deps]: Ok I just realized that we don't use the same one between api-tipi-build and nxxm-src

File: src/shipxx/shipxx.hxx (hunk 2 above)
Comment (Lambourl) [reply src/shipxx/shipxx.hxx:22]: log::debug() is an error, log class are only present in  nxxm-src

File: src/upgrd/upgrd.hxx [hunk 4]
```diff
@@ -206,17 +206,18 @@ namespace upgrd {
...
-                    auto str_cmd = "timeout /t 3 /nobreak & del /F /Q "s + _app_path.make_preferred().string() + " & "
-                      + "move /Y " + upgraded_app.make_preferred().string() + " " + _app_path.make_preferred().string();
-
-                    bp::spawn(system_shell, "/c", str_cmd.data());
-
+                    _log <<"Tipi may need administrative privileges to apply the update."<< std::endl;
+                    auto system_shell = bp::search_path("powershell.exe");
+                    auto str_cmd = "Start-Process cmd -ArgumentList \'timeout /t 3 /nobreak & del /F /Q "s + _app_path.make_preferred().string() + " & "
+                      + "move /Y " + upgraded_app.make_preferred().string() + " "s + _app_path.make_preferred().string(); +"\' -Verb RunAs";
+                    bp::spawn(system_shell,"-Command",str_cmd.data());
```
Comment (pysco68) [inline src/upgrd/upgrd.hxx]: That should not compile IMHO. There's a semi-colon after a string() call that's in the way.  If it does for some reason I don't understand the result wouldn't be the expected one.

Have you actually tried it?

`_app_path.make_preferred().string(); +"\' -Verb RunAs";`
` -----------------------------------^`

Comment (pysco68) [reply src/upgrd/upgrd.hxx]: + indenting and spaces. Btw. where's the style-guide you wanted to write?

**Review by pysco68 — CHANGES_REQUESTED** [review]
Please check

File: src/upgrd/upgrd.hxx (hunk 4 above)
Comment (Lambourl) [reply src/upgrd/upgrd.hxx]: Indeed it is a mistake on my part, I did not see it at all.
The style-guide is currently being written and I'm doing it in the background

Comment (daminetreg) [reply src/upgrd/upgrd.hxx]: > Indeed it is a mistake on my part, I did not see it at all. The style-guide is currently being written and I'm doing it in the background

Regarding a style guide I have a 5 bullet points one ready that exists that I can push and will accomodate flexibility and readability. And by the way we have a clang-format autoformat settings in nxxm, which is also a description of the expected style.

File: src/upgrd/upgrd.hxx [hunk 5]
```diff
@@ -214,7 +214,7 @@ namespace upgrd {
                     _log <<"Tipi may need administrative privileges to apply the update."<< std::endl;
                     auto system_shell = bp::search_path("powershell.exe");
                     auto str_cmd = "Start-Process cmd -ArgumentList \'timeout /t 3 /nobreak & del /F /Q "s + _app_path.make_preferred().string() + " & "
-                      + "move /Y " + upgraded_app.make_preferred().string() + " "s + _app_path.make_preferred().string(); +"\' -Verb RunAs";
+                                    + "move /Y " + upgraded_app.make_preferred().string() + " "s + _app_path.make_preferred().string() +"\' -Verb RunAs";
```
Comment (pysco68) [inline src/upgrd/upgrd.hxx]: Did you test that one now?

File: .nxxm/deps (hunk 1 above)
Comment (pysco68) [reply .nxxm/deps]: just on a side note, I think we have a version mismatch with kuba--/zip and my fork of his lib as well

File: src/upgrd/upgrd.hxx (hunk 5 above)
Comment (Lambourl) [reply src/upgrd/upgrd.hxx]: yes it is tested

File: .github/workflows/ci.yml [hunk 6]
```diff
@@ -4,13 +4,13 @@ on: [push]
...
+    container: tipibuild/tipi-ubuntu
     steps:
       - name: checkout
         uses: actions/checkout@v2
-      - name: nxxm builds project 
+      - name: tipi builds project 
         run: |
           export HOME=/root
-          nxxm . --verbose --test all 
+          tipi . -t linux-cxx17 --verbose
```
Comment (daminetreg) [inline .github/workflows/ci.yml:16]: No tests are run anymore here

File: src/shipxx/shipxx.hxx (hunk 2 above)
Comment (daminetreg) [reply src/shipxx/shipxx.hxx:22]: Still a problem we have ship duplicated, I thought we solved that issue

File: src/upgrd/upgrd.hxx [hunk 7]
```diff
@@ -206,17 +206,18 @@ namespace upgrd {
...
-
+                    _log << "Tipi may need administrative privileges to apply the update." << std::endl;
+                    auto system_shell = bp::search_path("powershell.exe");
+                    auto str_cmd = "Start-Process cmd -ArgumentList \'timeout /t 3 /nobreak & del /F /Q "s + _app_path.make_preferred().string() + " & "
+                                    + "move /Y " + upgraded_app.make_preferred().string() + " "s + _app_path.make_preferred().string() + "\' -Verb RunAs";
+                    bp::spawn(system_shell, "-Command", str_cmd.data());
                     #else 
+                    auto system_shell = bp::shell();
                     auto str_cmd = "sleep 3; rm "s + _app_path.generic_string() + ";"
                       + "mv " + upgraded_app.generic_string() + " " + _app_path.generic_string() + ";";
```
Comment (daminetreg) [inline src/upgrd/upgrd.hxx:223]: Linux should also request sudo right when needed.

**Review by daminetreg — CHANGES_REQUESTED** [review]
I think the issue with admin righ on linux/posix system is the same, we should request sudo rights to replace tipi during the upgrade.

- [ ] Request sudo right on linux if needed
- [ ] Why are the test disabled !?
- [ ] Factor out shipxx ( optional )

File: .github/workflows/ci.yml (hunk 6 above)
Comment (daminetreg) [reply .github/workflows/ci.yml:16]: I mean there is an example/ ok, I thought there were tests, so it seems we don't have any tests for the feature. We should add some to our e2e test suite.  Possibly we can turn the example in an unit test for this.

**Comment by daminetreg** [comment]
Probably dead now. Closing we can reopen it if we need it.

**Pull request closed without merging**